core/src/apps/cardano/auxiliary_data.py

```diff
@@ -0,0 +1,252 @@
+
+
+def get_auxiliary_data_bytes(
+    keychain: seed.Keychain,
+    auxiliary_data: Optional[CardanoTxAuxiliaryDataType],
+    protocol_magic: int,
+    network_id: int,
+) -> Optional[bytes]:
+    if not auxiliary_data:
+        return None
+
+    if auxiliary_data.type == CardanoAuxiliaryDataType.BLOB:
+        return auxiliary_data.blob
+    elif auxiliary_data.type == CardanoAuxiliaryDataType.TUPLE:
+        # ensured by _validate_auxiliary_data
+        assert auxiliary_data.metadata is not None
+        assert auxiliary_data.metadata.catalyst_registration_parameters is not None
+
+        cborized_catalyst_registration = _cborize_catalyst_registration(
+            keychain,
+            auxiliary_data.metadata.catalyst_registration_parameters,
+            protocol_magic,
+            network_id,
+        )
+
+        return cbor.encode(_wrap_metadata(cborized_catalyst_registration))
+    else:
+        raise INVALID_AUXILIARY_DATA
+
+
+def _cborize_catalyst_registration(
+    keychain: seed.Keychain,
+    catalyst_registration_parameters: CardanoCatalystRegistrationParametersType,
+    protocol_magic: int,
+    network_id: int,
+) -> CatalystRegistration:
+    staking_node = keychain.derive(catalyst_registration_parameters.staking_path)
+    staking_key = remove_ed25519_prefix(staking_node.public_key())
+
+    catalyst_registration_payload: CatalystRegistrationPayload = {
+        1: catalyst_registration_parameters.voting_public_key,
+        2: staking_key,
+        3: derive_address_bytes(
+            keychain,
+            catalyst_registration_parameters.reward_address_parameters,
+            protocol_magic,
+            network_id,
+        ),
+        4: catalyst_registration_parameters.nonce,
+    }
+
+    catalyst_registration_payload_signature = _sign_catalyst_registration_payload(
+        keychain,
+        catalyst_registration_payload,
+        catalyst_registration_parameters.staking_path,
+    )
+    catalyst_registration_signature = {1: catalyst_registration_payload_signature}
+
+    return {
+        CATALYST_REGISTRATION_KEY: catalyst_registration_payload,
+        CATALYST_REGISTRATION_SIGNATURE_KEY: catalyst_registration_signature,
+    }
+
+
+def _sign_catalyst_registration_payload(
+    keychain: seed.Keychain,
+    catalyst_registration_payload: CatalystRegistrationPayload,
+    path: List[int],
+) -> bytes:
+    node = keychain.derive(path)
+
+    encoded_catalyst_registration = cbor.encode(
+        {CATALYST_REGISTRATION_KEY: catalyst_registration_payload}
+    )
+
+    catalyst_registration_body_map_hash = hashlib.blake2b(
+        data=encoded_catalyst_registration,
+        outlen=VOTING_REGISTRATION_HASH_SIZE,
+    ).digest()
+
+    return ed25519.sign_ext(
+        node.private_key(), node.private_key_ext(), catalyst_registration_body_map_hash
+    )
+
+
+def _wrap_metadata(metadata: Dict) -> Tuple[Dict, Tuple]:
+    """
+    A new structure of metadata is used after Cardano Mary era. The metadata
```

Comment:
maybe worth including a link to cardano CDDL for reference: https://github.com/input-output-hk/cardano-ledger-specs/blob/f7deb22be14d31b535f56edc3ca542c548244c67/shelley-ma/shelley-ma-test/cddl-files/shelley-ma.cddl#L212

Reply:
https://github.com/vacuumlabs/trezor-firmware/pull/57/commits/39dddb339a5c4ace8e56f62818ee8a2330a123a2